geps/gep-1619.md

```diff
@@ -159,47 +167,380 @@ Generally, the implementation API programs the dataplane API; however these two
+    // 
+    // Support: Core
+    //
+    // +optional
+    Duration int64 `json:"duration,omitempty"`
+}
+
+
+// SessionPersistenceType defines a session persistence type.
+type SessionPersistenceType string
+
+const (
+    // SessionPersistenceTypeHttpCookie specifies to use an HTTP cookie for session persistence
+    SessionPersistenceTypeHttpCookie SessionPersistenceType = "HTTPCookie"
+)
+
+// SessionPersistencePolicyStatus defines the observed state of SessionPersistencePolicy.
+type SessionPersistencePolicyStatus struct {
+    // Conditions describe the current conditions of the SessionPersistencePolicy.
+    //
+    // Implementations should prefer to express SessionPersistencePolicy
+    // conditions using the `SessionPersistencePolicyConditionType` and
+    // `SessionPersistencePolicyConditionReason` constants so that
+    // operators and tools can converge on a common vocabulary to
+    // describe SessionPersistencePolicy state.
+    // Known condition types are:
+    //
+    // * “Accepted”
+    //
+    // +optional
+    // +listType=map
+    // +listMapKey=type
+    // +kubebuilder:validation:MaxItems=8
+    // +kubebuilder:default={type: "Accepted", status: "Unknown", reason:"Pending", message:"Waiting for validation", lastTransitionTime: "1970-01-01T00:00:00Z"}
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// SessionPersistencePolicyConditionType is the type of condition used
+// as a signal by SessionPersistencePolicy. This type should be used with
+// the SessionPersistencePolicy.Conditions field.
+type SessionPersistencePolicyConditionType string
+
+// SessionPersistencePolicyConditionReason is a reason that explains why a
+// particular SessionPersistencePolicyConditionType was generated. This reason
+// should be used with the SessionPersistencePolicy.Conditions field.
+type SessionPersistencePolicyConditionReason string
+
+const (
+    // This condition indicates that the SessionPersistencePolicyStatus has been
+    // accepted as valid.
+    // Possible reason for this condition to be True is:
+    //
+    // * “Accepted”
+    //
+    // Possible reasons for this condition to be False are:
+    //
+    // * “Invalid”
+    // * “Pending”
+    SessionPersistencePolicyConditionAccepted SessionPersistencePolicyConditionType = “Accepted”
+
+    // This reason is used with the “Accepted” condition when the condition is true.
+    SessionPersistencePolicyReasonAccepted SessionPersistencePolicyConditionReason = “Valid”
+
+	// This reason is used with the “Accepted” condition when the SessionPersistencePolicy is invalid, e.g. crossing namespace boundaries.
+    SessionPersistencePolicyReasonInvalid SessionPersistencePolicyConditionReason = “Invalid”
+
+    // This reason is used with the “Accepted” condition when the SessionPersistencePolicy is pending validation.
+    SessionPersistencePolicyReasonPending SessionPersistencePolicyConditionReason = “Pending”
+)
+```
 
 ### Session Affinity API
```

Comment:
I have been thinking about this quite a bit, and am a bit uncomfortable with the ambiguities and "foot guns" we are exposing with this API.

The GEP as-is basically has no details on how persistence will work, really just "This thing should be sticky". While I think this would be fine for trivial cases, it ends up being extremely complex in non trivial cases.

There are a variety of cases that are Hard to solve. Just a few selected ones of the worst case scenarios, where I don't know _any_ impementation that handles these well (there are tons of other cases that some implementations handle poorly).

**Multiple services with affinity**

Consider a service A and B, both with affinity configured.

Given all existing implementations use a single cookie per service, or allow users to configure the cookie explicitly (and therefore, also use a single cookie per service), this has the potential to have a broken setup.

A trivial curl A; curl B is fine due to cookies being keyed by Host (and, optionally Path). However, there are many cases where these mitigations do not work.

For example:
```
Route:
- match: header foo=bar
  backends:
  - A
- match:
  backends:
  - B
```
In this case, affinity is lost each time we switch from foo=bar being present or not.

Even path cases may be problematic, as this doesn't handle all match types (Regex, for example). The GEP kind of handles regex, but not in a reliable way.


**Traffic splitting**

The GEP talks about this, but its pretty vague: " When session persistence is combined with traffic splitting, users might desire the session persistence to take precedence over the traffic splitting pattern. In such situations, once a persistent session is established, the traffic should no longer be split, but should instead be consistently directed to the same backend." Basically "whatever the implementation decides is valid".

Very similar to the above, where we are explicitly splitting services by request contents, we can also implicitly do it with weighting.

Some implementations simply break here and lose affinity each time a different backend is selected. 

Others prioritize the affinity over the traffic split. However, this has interesting implications. 

Consider the below route:
```
route:
- match: /a
  backends: v1
- match: /b
  backends:
  - v1, weight=0
  - v2, weight=100
```

In this case, curl app/a; curl app/b will send requests to v1. This should be impossible by the spec (weight=0).

----

Overall, there are a ton of cases where a user my configure Affinity for <Thing>, and it works. Then they make a seemingly unrelated change, and affinity suddenly stops working with no signals.

I think we need to be a lot more prescriptive here. Some options:
* Be extremely strict on where we allow affinity to be applied to avoid these cases
* Require that if users get into a "risky configuration" a status is reported on the affinity
* Explicitly document all the ways affinity can be broken. Or the inverse, document the only allowed configurations for affinity
* Be more prescriptive of exactly how implementations should handle these cases.

Reply:
Another example of 'impossible' behavior (this one is fairly implementation specific, but is how gRPC works and how Envoy is proposed to work):

* svc A with affinity
* svc B without affinity 
* route Y without affinity with backends A, B

At this point, traffic to B never uses affinity.

Now, I add a new route:
* route X, which has `match` that never occurs, with affinity with backends ends A, B

Traffic through route Y to backend B will now incorrectly use affinity even though it is not configured to do so, because of a completely different route, which cannot ever match.

Reply:
@howardjohn I think we discussed this offline. As mentioned before this is a made-up example and what a user will/should do in practice is as follows:

- remove route X since it never matches so it is not serving any purpose 
- split route Y into 2 routes: Y1 for svc A and Y2 for svc B and attach affinity to Y1.

Reply:
> @howardjohn I think we discussed this offline. As mentioned before this is a made-up example and what a user will/should do in practice is as follows:
> 
> * remove route X since it never matches so it is not serving any purpose
> * split route Y into 2 routes: Y1 for svc A and Y2 for svc B and attach affinity to Y1.

I don't agree...

> remove route X since it never matches so it is not serving any purpose

The "Route X which never matches" was a contrived extreme example - the same issues occurs if X matches.. at which point they cannot remove it. 

> split route Y into 2 routes: Y1 for svc A and Y2 for svc B and attach affinity to Y1.

This is completely different behavior, you cannot always just split routes  - that would then be a different config. For example the route can have a traffic split between A and B in Y -- that cannot be split into two routes

> what a user will/should do in practice

In practice, they will probably slam their keyboard, rage on twitter, and never use the product again because two completely unrelated configurations somehow broke each other - something that should never happen. This is not hypothetical - Istio has many many APIs with the same issue and the results are usually something like this. Gateway API, thus far, has avoided making the same mistakes.

I doubt any user will apply Route X and realize that this route -- which is completely related from Route Y, for a different domain, probably owned by another namespace and team -- somehow broke Route Y.

Reply:
@howardjohn Thanks for concrete examples. I agree this PR currently lacks details on the expected behavior. I think I've been conflating @costinm's comments on "don't specify how Session Persistence is accomplished" with "defining an expected behavior". We can specify an expected behavior without specifying how exactly it's accomplished.

Could you help me define...defining an expected behavior....further? I think it could be broken down into:
1. Defining the session persistence behavior as it pertains to API spec configuration (i.e. input config and output behavior)
2. Defining the session persistence behavior as it interacts with components in your cluster

I believe what you have been referring to here is # 1: defining the behavior when a user applies a specific Gateway API spec configuration to enable session persistence (for example, what happens when you attach persistence to a route vs. service). I agree, these specific API spec configuration use cases and their need to be defined. I'll work on defining these cases.

However, for # 2, I am not so clear. An example would be the scenario of "Cookie Rewriting". Another component (backend) is injecting a cookie with the same name, in which an implementation needs to decide on how to handle. This scenario wasn't created by an user's API configuration like # 1, but instead is circumstantial. Is this just an implementation detail? What if the implementation's design decision results in no session persistence because there's already a cookie with the same name? Couldn't this be a "slam the keyboard" moment too?

Reply:
FYI added [`Expected API Behavior`](https://github.com/kubernetes-sigs/gateway-api/pull/2159/files#diff-afa0e827962fe48152635217c6bc954020d5fec3578af4a9e5f744a2b4568fffR533) section as solution to John's suggestion: `Be more prescriptive of exactly how implementations should handle these cases.`

I also am starting to wonder if we are biting off too much by attempting to apply session persistence to a `route` in this revision of the GEP. Costin has been warning us that it's complicated and confusing for implementations and I think we are introducing a lot of edge cases. Attaching to services seem much simpler. We can keep trying for route attachments in this revision, but if it gets too messy, we could also postpone. Thoughts?

Reply:
Yea I tend to agree with Costin's comment about this, the hard part seems to be attaching to routes, where services seem to be more common of a use case:

> Not sure how we can do the reverse - route attachment is where the
complexity and strange cases lie.
> Per service - as used in several products already, including K8S Service -
may have some corner cases, but it is quite common.

Though, the hardest issues arise when combing them both.